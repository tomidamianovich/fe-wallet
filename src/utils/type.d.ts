export type CurrencyType = {
  ticker: string
  name: string
  symbol: string
  decimals: number
  type: string
  actions: {
    action: string
    enabled: boolean
  }[]
  url_images: {
    image_png: string
    image_svg: string
  }
  color: string | null
  categories: string[]
  is_favorite: boolean
}

export type RateType = {
  ticker: string
  buy_rate: string
  sell_rate: string
  variation: string
}

export type BalanceType = {
  ticker: string
  amount: number
  account_limit: number
}

export type TransactionActionType =
  | 'DEPOSIT'
  | 'WITHDRAWAL'
  | 'SEND'
  | 'RECEIVE'
  | 'BUY'
  | 'SWAP'
export type TransactionStatusType = 'PENDING' | 'FAILURE' | 'SUCCESS'

export type TransactionType = {
  date: Date
  action: TransactionActionType
  ticker: string
  via: string
  status: TransactionStatusType
  comision: number
  amount: number
  total: number
}

export interface transaction {
  ['DEPOSIT']: TransactionActionType
  ['WITHDRAWAL']: TransactionActionType
  ['SEND']: TransactionActionType
  ['RECEIVE']: TransactionActionType
  ['BUY']: TransactionActionType
  ['SWAP']: TransactionActionType
}
export interface transactionTitle {
  DEPOSIT: string
  WITHDRAWAL: string
  SEND: string
  RECEIVE: string
  BUY: string
  SWAP: string
}

export interface transactionStatus {
  ['PENDING']: TransactionStatusType
  ['FAILURE']: TransactionStatusType
  ['SUCCESS']: TransactionStatusType
}

export type Link = {
  name: string
  to: string
  icon: string
}

export type EventSelect = {
  target: {
    value: TransactionActionType
  }
}

export type NavLinkType = link

export type CurrencyState = CurrencyType[]
export type RateState = RateType[]
export type NavLinkState = NavLinkType[]
export type BalanceState = BalanceType[]
export type TransactionState = TransactionType[]

export type CurrencyAction = {
  type: strings
  payload: CurrencyState
}

export type RateAction = {
  type: string
  payload: RateState
}

export type NavLinkAction = {
  type: string
  payload: NavLinkState
}

export type BalanceAction = {
  type: string
  payload: BalanceState
}

export type CurrentBalanceAction = {
  type: string
  payload: BalanceType
}

export type TransactionsAction = {
  type: string
  payload: TransactionState
}

export type BalanceWithCurrencyInfo = BalanceType &
  CurrencyType & { sell_rate: string }

export type CurrencyDispatchType = (args: CurrencyAction) => CurrencyAction
export type RateDispatchType = (args: RateAction) => RateAction
export type NavLinkDispatchType = (args: RateAction) => NavLinkAction
export type BalanceDispatchType = (args: BalanceAction) => BalanceAction
export type TransactionDispatchType = (
  args: TransactionsAction,
) => TransactionsAction

// Combined Types

export type CombinedState = {
  CurrencyReducer: CurrencyState
  RateReducer: RateState
  NavLinkReducer: NavLinkState
  BalanceReducer: BalanceState
  CurrentBalanceReducer: BalanceType
  TransactionsReducer: TransactionState
}

export type CombinedAction = {
  CurrencyReducer: CurrencyAction
  RateReducer: RateAction
  NavLinkReducer: NavLinkAction
  BalanceReducer: BalanceAction
  CurrentBalanceReducer: CurrentBalanceAction
  TransactionsReducer: TransactionsAction
}

export type CombinedDispatchType = (args: CombinedAction) => CombinedAction
