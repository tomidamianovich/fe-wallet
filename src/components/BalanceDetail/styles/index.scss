.fe-wallet__balance-detail {
  display: flex;
  flex-wrap: wrap;
  justify-content: flex-start;
  align-content: center;
  align-items: center;
  flex-direction: row;
  margin-top: 32px;
  width: calc(100% + 32px);

  &__card {
    margin: 0 32px 16px 0;
    height: 160px;
    min-width: 352px;
    left: 0px;
    top: 0px;
    border-radius: 6px;
    background: #FFFFFF;
    flex-grow: 0;
    flex-shrink: 0;
    display: flex;
    flex-basis: calc(33.33% - 32px);
    align-items: center;
    justify-content: space-evenly;
    flex-wrap: wrap;
    flex-direction: column;

    &__header {
      display: flex;
      width: 100%;
      justify-content: space-between;
      box-sizing: border-box;
      padding: 0 16px;
      &__title {
        display: flex;
        justify-content: space-between;
        color: #A5A7AF;
        font-size: 16px;
        font-weight: 400;
        line-height: 19px;
        letter-spacing: 0px;
        text-align: left;
      }
      &__buy {
        background: none;
        color: #7908FF;
        font-size: 16px;
        border: none;
        padding: 0;
        font: inherit;
        cursor: pointer;
        outline: inherit;
        display: flex;
        align-items: center;
        .icon {
          padding-left: 5px;
          path {
            fill: #7908FF;
          }
        }
      }
    }

    &__info {
      display: flex;
      justify-content: center;
      align-items: center;
      &__values {
        color: #373744;
        font-size: 32px;
        font-weight: 500;
        line-height: 32px;
        letter-spacing: 0px;
        text-align: center;
      }

      img {
        margin-right: 6px;
      }
    }

    &__current-currency {
      color: #A5A7AF;
      font-size: 16px;
      font-weight: 400;
      line-height: 20px;
      letter-spacing: 0px;
      text-align: center;
    }
  }
}