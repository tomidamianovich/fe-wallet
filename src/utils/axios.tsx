import { RateType, CurrencyType } from '../utils/type';
const axios = require('axios').default;

type responseType = {
  data: RateType[] | CurrencyType[]
}
  
const requestHandler = async (url:string): Promise<any> => 
  new Promise((resolve, reject) => 
    axios.get(url)
    .then((response: responseType) => resolve(response?.data))
    .catch((err:any) => reject(err))
  );

export {
  requestHandler,
}