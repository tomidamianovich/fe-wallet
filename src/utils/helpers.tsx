const handleFormatValues = (value:number) => value.toFixed(2);

export {
  handleFormatValues
}