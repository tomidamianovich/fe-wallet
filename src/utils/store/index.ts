import currencies from './currencies';
import cryptoCurrencies from './cryptoCurrencies';
import fiatCurrencies from './fiatCurrencies';
import balances from './balances';
import rates from './rates';

export {
  currencies,
  fiatCurrencies,
  cryptoCurrencies,
  rates,
  balances
}