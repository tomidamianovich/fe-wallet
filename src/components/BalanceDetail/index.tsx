import React, { useState } from 'react';
import { useEffect } from 'react';
import { BalanceType, CurrencyType, RateType, CombinedState } from '../../utils/type';
import { useSelector } from 'react-redux';
import BalanceDetailCard from '../BalanceDetailCard';
import './styles/index.scss';

type Props = {
  balances: BalanceType[],
  currencies: CurrencyType[],
  rates: RateType[]
};

type BalanceWithCurrencyInfo = BalanceType & CurrencyType & { sell_rate: string } ;

const BalanceDetail: React.FC<Props> = ({
  balances,
  currencies,
  rates
}) => {
  const currentBalance:BalanceType  = useSelector((state:CombinedState) => state.CurrentBalanceReducer);
  const [cryptosBalance, setCryptosBalance] = useState<BalanceWithCurrencyInfo[]>([]);

  useEffect(()=> {
    const cryptoTickers = currencies.map(currency => currency.ticker);
    const cryptosBalance = balances.reduce((acc:BalanceWithCurrencyInfo[], val: BalanceType) => {
      if (!(cryptoTickers.includes(val.ticker) && !!val.amount)) return acc
      const currencyData:CurrencyType = currencies.filter(currency => currency.ticker === val.ticker)[0];
      const { sell_rate } = rates.find(rate => rate.ticker === `${val.ticker}_${currentBalance?.ticker}`) ?? {sell_rate: "0"};
      acc.push({
        ...val, 
        ...currencyData,
        sell_rate
      });
      return acc 
    }, []);
    setCryptosBalance(cryptosBalance);
  }, [setCryptosBalance, balances, currencies, rates, currentBalance]);

  return(
    <div className="fe-wallet__balance-detail">
      { cryptosBalance.map(balance => <BalanceDetailCard {...balance} key={balance.ticker} />) }
    </div>
  )
};

export default BalanceDetail;