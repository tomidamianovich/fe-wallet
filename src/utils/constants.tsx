import { transaction, transactionTitle, transactionStatus } from '../utils/type';

const ENDPOINTS = {
  CURRENCIES: 'http://localhost:3001/currencies',
  RATES: 'http://localhost:3001/rates',
  BALANCES: 'http://localhost:3001/balance'
}

const PANEL = {
  HEADING: "Mi billetera",
  SUBHEADING: "En tu billetera vas a poder almacenar todas las criptomonedas que compres en ripio.",
}

const TRANSACTIONS:{
  HEADING: string,
  HEADER: string[],
  TYPES: transaction,
  TYPES_TITLES: transactionTitle,
  STATUS: transactionStatus
} = {
  HEADING: "Mis movimientos",
  HEADER: [
    "Fecha",
    "Tipo de Operación",
    "Método",
    "Estado",
    "Comisión",
    "Monto",
  ],
  TYPES: {
    DEPOSIT: "DEPOSIT",
    WITHDRAWAL: "WITHDRAWAL",
    SEND: "SEND",
    RECEIVE: "RECEIVE",
    BUY: "BUY",
    SWAP: "SWAP"
  },
  TYPES_TITLES: {
    DEPOSIT: "Ingreso",
    WITHDRAWAL: "Retiro",
    SEND: "Envio",
    RECEIVE: "Recibo",
    BUY: "Compra",
    SWAP: "Swap"
  },
  STATUS: {
    PENDING: "PENDING",
    FAILURE: "FAILURE",
    SUCCESS: "SUCCESS"
  }
}

const DEFAULT_BALANCE = {
  ticker: "",
  amount: 0,
  account_limit: 0,
  name: "",
  symbol: "",
  decimals: 2,
  type: "",
  actions: [{
    action: "",
    enabled: true
  }],
  url_images: {
    image_png: "",
    image_svg: "",
  },
  color: "",
  categories: [],
  is_favorite: true,
  sell_rate: ""
}

const COMISIONS = {
  BUY: 0.003,
  DEPOSIT: 0,
  WITHDRAWAL: 0.0004,
  SEND: 0,
  RECEIVE: 0,
  SWAP: 0.00004
}

type Link = {
  name: string;
  to: string;
  icon: string;
};

const NAV_LINKS:Link[] = [
  {
    name: 'Panel',
    to: 'panel',
    icon: 'panel'
  },
  {
    name: 'Billetera',
    to: 'wallet',
    icon: 'wallet'
  },
  {
    name: 'Cotizaciones',
    to: 'market-rates',
    icon: 'market-rates'
  },
  {
    name: 'Créditos',
    to: 'credit',
    icon: 'credit'
  },
  {
    name: 'Exchange',
    to: 'exchange',
    icon: 'exchange'
  },
  {
    name: 'Launchpad',
    to: 'launchpad',
    icon: 'launchpad'
  }
];

const WORDINGS = {
  BALANCE: {
    DETAIL: {
      INPUT: {
        LABELS: {
          TRANSACTION: "Seleccione el tipo de operacion a realizar",
          AMOUNT: "Ingrese cantidad",
          ARIA_LABEL : "Monto de la transacción"
        },
        MESSAGES: {
          INSUFFICIENT_AMOUNT: "Dinero insuficiente",
          AMOUNT_TO_DEBIT: "Monto a debitar",
          AMOUNT_NEEDED: "Monto necesario",
          ACTUAL_BALANCE: "Monto actual"
        }
      },
      ACTIONS: {
        CANCEL: {
          LABEL: "Cancelar",
          ARIA_LABEL: "Cancelar transacción",
        },
        CONFIRM: {
          LABEL: "Confirmar",
          ARIA_LABEL: "Confirmar transacción",
        }
      },
      CARD: {
        ACCOUNT: "Cuenta"
      }
    },
    SUMMARY: {
      DROPDOWN: {
        ARIA_LABEL: "Seleccionar moneda para ver balance"
      }
    }
  },
  CONVERTER: {
    FROM: {
      DROPDOWN: {
        ARIA_LABEL: "Seleccionar crypto origen",
        NAME: "from"
      },
      INPUT: {
        ARIA_LABEL: "Cripto a convertir",
        NAME: "from_amount"
      }
    },
    TO: {
      DROPDOWN: {
        ARIA_LABEL: "Seleccionar crypto destino",
        NAME: "to"
      },
      INPUT: {
        ARIA_LABEL: "Cripto convertida",
        NAME: "to_amount"
      }
    },
    BUTTON: {
      ARIA_LABEL: "Convertir",
      LABEL: "Convertir",
    }
  },
  CURRENCIES: {
    TYPES: {
      FIAT: "FIAT",
      CRYPTO: "CRYPTO"
    }
  },
  MAINTENANCE: {
    TITLE: "Pagina en Mantenimiento",
    SUBTITLE: {
      BEGINNING: "Oops! Estamos trabajando en la pagina ",
      ENDING: ". Ingrese nuevamente a esta seccion más tarde."
    }
  },
  NAVIGATION: {
    NAV_ITEM: {
      NAME: "Salir"
    }
  },
  NOT_FOUND: "Pagina no encontrada.",
  ERROR_FOUND: "Oops. Ocurrio un error",
  PAGES: {
    WALLET: "wallet",
    CREDIT: "credit",
    EXCHANGE: "exchange",
    LAUNCHPAD: "launchpad",
    MARKET_RATES: "market-rates",
    PANEL: "panel",
  }
}

export {
  ENDPOINTS,
  PANEL,
  TRANSACTIONS,
  DEFAULT_BALANCE,
  COMISIONS,
  NAV_LINKS,
  WORDINGS
}