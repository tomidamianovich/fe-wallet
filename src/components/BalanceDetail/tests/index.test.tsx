import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import BalanceDetail from '../index';
import { balances, fiatCurrencies, currencies, rates } from '../../../utils/tests';
import { createStore, applyMiddleware } from "redux";
import { Provider } from 'react-redux';

const store = createStore(()=>[], applyMiddleware());

test('renders BalanceDetail with balances', () => {
  const { asFragment } = render(
    <Provider store={store}>
      <BalanceDetail 
        balances={balances}
        currencies={currencies}
        rates={rates}
      />
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});