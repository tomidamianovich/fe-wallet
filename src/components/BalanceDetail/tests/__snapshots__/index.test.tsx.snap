// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders BalanceDetail with balances 1`] = `
<DocumentFragment>
  <div
    class="fe-wallet__balance-detail"
  >
    <div
      class="fe-wallet__balance-detail__card"
    >
      <div
        class="fe-wallet__balance-detail__card__header"
      >
        <span
          class="fe-wallet__balance-detail__card__header__title"
        >
          Cuenta Pesos
        </span>
        <button
          class="fe-wallet__balance-detail__card__header__buy"
        >
          Comprar
          <svg
            class="icon"
            height="15"
            viewBox="0 0 16 15"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M8.2 2C8.86274 2 9.4 2.53726 9.4 3.2C9.4 3.86274 8.86274 4.4 8.2 4.4C7.53726 4.4 7 3.86274 7 3.2C7 2.53726 7.53726 2 8.2 2ZM9.4 8C9.4 7.33726 8.86274 6.8 8.2 6.8C7.53726 6.8 7 7.33726 7 8C7 8.66274 7.53726 9.2 8.2 9.2C8.86274 9.2 9.4 8.66274 9.4 8ZM9.4 12.8C9.4 12.1373 8.86274 11.6 8.2 11.6C7.53726 11.6 7 12.1373 7 12.8C7 13.4627 7.53726 14 8.2 14C8.86274 14 9.4 13.4627 9.4 12.8Z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
      <div
        class="fe-wallet__balance-detail__card__info"
      >
        <img
          alt="Pesos"
          src="https://cmswagtail.s3.sa-east-1.amazonaws.com/crypto-assets/SVG/ARS.svg"
        />
        <span
          class="fe-wallet__balance-detail__card__info__values"
        >
          98000 $
        </span>
      </div>
      <div
        class="fe-wallet__balance-detail__card__current-currency"
      >
        <span>
          ≈ 0.00 undefined
        </span>
      </div>
    </div>
    <div
      class="fe-wallet__balance-detail__card"
    >
      <div
        class="fe-wallet__balance-detail__card__header"
      >
        <span
          class="fe-wallet__balance-detail__card__header__title"
        >
          Cuenta Bitcoin
        </span>
        <button
          class="fe-wallet__balance-detail__card__header__buy"
        >
          Comprar
          <svg
            class="icon"
            height="15"
            viewBox="0 0 16 15"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M8.2 2C8.86274 2 9.4 2.53726 9.4 3.2C9.4 3.86274 8.86274 4.4 8.2 4.4C7.53726 4.4 7 3.86274 7 3.2C7 2.53726 7.53726 2 8.2 2ZM9.4 8C9.4 7.33726 8.86274 6.8 8.2 6.8C7.53726 6.8 7 7.33726 7 8C7 8.66274 7.53726 9.2 8.2 9.2C8.86274 9.2 9.4 8.66274 9.4 8ZM9.4 12.8C9.4 12.1373 8.86274 11.6 8.2 11.6C7.53726 11.6 7 12.1373 7 12.8C7 13.4627 7.53726 14 8.2 14C8.86274 14 9.4 13.4627 9.4 12.8Z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
      <div
        class="fe-wallet__balance-detail__card__info"
      >
        <img
          alt="Bitcoin"
          src="https://cmswagtail.s3.sa-east-1.amazonaws.com/crypto-assets/SVG/BTC.svg"
        />
        <span
          class="fe-wallet__balance-detail__card__info__values"
        >
          0.00948 BTC
        </span>
      </div>
      <div
        class="fe-wallet__balance-detail__card__current-currency"
      >
        <span>
          ≈ 0.00 undefined
        </span>
      </div>
    </div>
    <div
      class="fe-wallet__balance-detail__card"
    >
      <div
        class="fe-wallet__balance-detail__card__header"
      >
        <span
          class="fe-wallet__balance-detail__card__header__title"
        >
          Cuenta Ethereum
        </span>
        <button
          class="fe-wallet__balance-detail__card__header__buy"
        >
          Comprar
          <svg
            class="icon"
            height="15"
            viewBox="0 0 16 15"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M8.2 2C8.86274 2 9.4 2.53726 9.4 3.2C9.4 3.86274 8.86274 4.4 8.2 4.4C7.53726 4.4 7 3.86274 7 3.2C7 2.53726 7.53726 2 8.2 2ZM9.4 8C9.4 7.33726 8.86274 6.8 8.2 6.8C7.53726 6.8 7 7.33726 7 8C7 8.66274 7.53726 9.2 8.2 9.2C8.86274 9.2 9.4 8.66274 9.4 8ZM9.4 12.8C9.4 12.1373 8.86274 11.6 8.2 11.6C7.53726 11.6 7 12.1373 7 12.8C7 13.4627 7.53726 14 8.2 14C8.86274 14 9.4 13.4627 9.4 12.8Z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
      <div
        class="fe-wallet__balance-detail__card__info"
      >
        <img
          alt="Ethereum"
          src="https://cmswagtail.s3.sa-east-1.amazonaws.com/crypto-assets/SVG/ETH.svg"
        />
        <span
          class="fe-wallet__balance-detail__card__info__values"
        >
          1.23456 ETH
        </span>
      </div>
      <div
        class="fe-wallet__balance-detail__card__current-currency"
      >
        <span>
          ≈ 0.00 undefined
        </span>
      </div>
    </div>
    <div
      class="fe-wallet__balance-detail__card"
    >
      <div
        class="fe-wallet__balance-detail__card__header"
      >
        <span
          class="fe-wallet__balance-detail__card__header__title"
        >
          Cuenta Litecoin
        </span>
        <button
          class="fe-wallet__balance-detail__card__header__buy"
        >
          Comprar
          <svg
            class="icon"
            height="15"
            viewBox="0 0 16 15"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M8.2 2C8.86274 2 9.4 2.53726 9.4 3.2C9.4 3.86274 8.86274 4.4 8.2 4.4C7.53726 4.4 7 3.86274 7 3.2C7 2.53726 7.53726 2 8.2 2ZM9.4 8C9.4 7.33726 8.86274 6.8 8.2 6.8C7.53726 6.8 7 7.33726 7 8C7 8.66274 7.53726 9.2 8.2 9.2C8.86274 9.2 9.4 8.66274 9.4 8ZM9.4 12.8C9.4 12.1373 8.86274 11.6 8.2 11.6C7.53726 11.6 7 12.1373 7 12.8C7 13.4627 7.53726 14 8.2 14C8.86274 14 9.4 13.4627 9.4 12.8Z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
      <div
        class="fe-wallet__balance-detail__card__info"
      >
        <img
          alt="Litecoin"
          src="https://cmswagtail.s3.sa-east-1.amazonaws.com/crypto-assets/SVG/LTC.svg"
        />
        <span
          class="fe-wallet__balance-detail__card__info__values"
        >
          1.23456 LTC
        </span>
      </div>
      <div
        class="fe-wallet__balance-detail__card__current-currency"
      >
        <span>
          ≈ 0.00 undefined
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
